File: YYReader/Support/ReaderPageScroll.swift
enum ReaderPageScroll {
    static let pageFraction = 0.88
    static let pageOverlap = 1 - pageFraction
    static let smallStep = 112.0
    static let animationDuration = 0.15

    static func pageDistance(viewportHeight: Double) -> Double {
        max(viewportHeight, 0) * pageFraction
    }

    static func pageDestinationY(
        currentY: Double,
        viewportHeight: Double,
        contentHeight: Double,
        direction: Double
    ) -> Double {
        destinationY(
            currentY: currentY,
            viewportHeight: viewportHeight,
            contentHeight: contentHeight,
            distance: pageDistance(viewportHeight: viewportHeight) * direction
        )
    }

    static func destinationY(
        currentY: Double,
        viewportHeight: Double,
        contentHeight: Double,
        distance: Double
    ) -> Double {
        let maximumY = max(contentHeight - viewportHeight, 0)
        return min(max(currentY + distance, 0), maximumY)
    }

    static func shouldAnimate(reduceMotion: Bool, isKeyRepeat: Bool = false) -> Bool {
        !reduceMotion && !isKeyRepeat
    }
}

File: windows/YYReader.Windows.Core/Models/ReaderPageScroll.cs
namespace YYReader.Windows.Core.Models;

public static class ReaderPageScroll
{
    public const double PageFraction = 0.88;
    public const double SmallStep = 112;

    public static double PageDistance(double viewportHeight) => Math.Max(viewportHeight, 0) * PageFraction;

    public static bool ShouldLoadNext(double verticalOffset, double scrollableHeight, double viewportHeight)
    {
        if (viewportHeight <= 0)
        {
            return false;
        }

        var remainingDistance = Math.Max(scrollableHeight - verticalOffset, 0);
        var preloadDistance = Math.Max(viewportHeight * 1.25, 480);
        return remainingDistance <= preloadDistance;
    }

    public static double DestinationY(double currentY, double viewportHeight, double contentHeight, double distance)
    {
        var maximumY = Math.Max(contentHeight - viewportHeight, 0);
        return Math.Clamp(currentY + distance, 0, maximumY);
    }
}
